### PR TITLE
Bug 1949075: Add AddPage customization object

### PR DESCRIPTION
**Pull request by jerolimov**
**Issues:**
Epic: https://issues.redhat.com/browse/ODC-5013
Story: https://issues.redhat.com/browse/ODC-5416
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1949075

Add AddPage customization API based on https://github.com/openshift/enhancements/pull/669

## Test setup (API change only with console-operator PR)

This change could be tested with https://github.com/openshift/console-operator/pull/527

1. Launch a cluster with `launch openshift/console-operator#527`

2. Open `https://cluster/k8s/cluster/operator.openshift.io~v1~Console/cluster/yaml` and configure a customization like:

```yaml
spec:
  customization:
    addPage:
      disabledActions:
        - git
        - git2
        - import-from-samples
```

3. Open `https://cluster/k8s/ns/openshift-console/configmaps/console-config` and check data key `console-config.yaml`. The YAML should contain the `disabledActions` after a few seconds.

```yaml
customization:
  addPage:
    disabledActions:
    - git
    - git2
    - import-from-samples
```

## Test setup E2E incl. frontend change

To test this change together with https://github.com/openshift/console-operator/pull/527 and https://github.com/openshift/console/pull/8643 you need to start cluster with the console-operator PR and the frontend change:

1. Launch a cluster with `launch openshift/console-operator#527,openshift/console#8643`

2. Search for "Console", select apiVersion "operator.openshift.io/v1" and select the "cluster" resource.

3. Open the YAML tab (Quick link for this resource: `https://cluster/k8s/cluster/operator.openshift.io~v1~Console/cluster/yaml`)

4. Add customization:

```yaml
spec:
  customization:
    addPage:
      disabledActions:
        - git
        - git2
        - import-from-samples
```

5. Alternative you can just add `spec.customization.addPage.disabledActions` (see above) as YAML and then insert all add action IDs from the snippet sidebar.

6. Open `https://cluster/k8s/ns/openshift-console/configmaps/console-config` and check data key `console-config.yaml`. The YAML should contain the `disabledActions` after a few seconds.

```yaml
customization:
  addPage:
    disabledActions:
    - git
    - git2
    - import-from-samples
```

7. You need to reload your browser to get the latest customization loaded. Unfortunately they are not loaded directly. So please remember to reload the page if your latest change are not displayed.

8. Open the Developer perspective, and open the Add page. Disabled add actions should now not displayed anymore.

**Full demo:** https://user-images.githubusercontent.com/139310/116062842-932c9b00-a684-11eb-91b9-d1279fb954f8.mp4

### CODE REVIEW

File: operator/v1/types_console.go
Comment (jhadvig) [inline operator/v1/types_console.go:126]:
```suggestion
	// addPage allows customizing actions on the Add page in developer perspective.
```

File: operator/v1/types_console.go
Comment (jhadvig) [inline operator/v1/types_console.go:190]:
```suggestion
// AddPage allows customizing actions on the Add page in developer perspective.
```

File: operator/v1/types_console.go
Comment (jhadvig) [inline operator/v1/types_console.go:192]:
```suggestion
	// disabledActions is a list of actions that are not shown to users. Each action in the list is represented by it's ID.
```

File: operator/v1/types_console.go
Comment (jhadvig) [inline operator/v1/types_console.go:127]: Lets add:
`// +kubebuilder:validation:MinItems=1`

**Review by jhadvig — CHANGES_REQUESTED** [review]
@jerolimov thanks for the PR, looks good.
Two wording nits and lets add an additional validation.
Otherwise looks 👌

File: operator/v1/types_console.go
Comment (jerolimov) [reply operator/v1/types_console.go:192]: Okay. I change it to 

```
	// disabledActions is a list of actions that are not shown to users.
	// Each action in the list is represented by its ID.
```

**Comment by jerolimov** [comment]
@jhadvig Thanks for this wording hints. Applied all your change requests.

I will test the additional validation also on a cluster to verify if this works as expected.

**Comment by jhadvig** [comment]
> I still could save disablesActions: [""] (an empty string), but I think this is "okayish" or do you now an additional kubebuilder:validation rule for the string array?

Not sure if in this case we could use, to specify the length of a string instance:
`// +kubebuilder:validation:MinLength=1`

In addition I would also add:
`// +kubebuilder:validation:UniqueItems`

**Comment by jerolimov** [comment]
@jhadvig unfortunately both additional attributes doesn't work:

`// +kubebuilder:validation:MinLength=1` doesn't work for `[]string`:

```
$ make update build verify

Running `gofmt -s -l -w` on 287 file(s).
Using existing controller-gen from "_output/tools/bin/controller-gen"
Using existing yq from "_output/tools/bin/yq"
Using existing yaml-patch from "_output/tools/bin/yaml-patch"
'_output/tools/bin/controller-gen' schemapatch:manifests="./authorization/v1" paths="./authorization/v1" output:dir="./authorization/v1"
'_output/tools/bin/controller-gen' schemapatch:manifests="./apiserver/v1" paths="./apiserver/v1" output:dir="./apiserver/v1"
'_output/tools/bin/controller-gen' schemapatch:manifests="./config/v1" paths="./config/v1" output:dir="./config/v1"
'_output/tools/bin/controller-gen' schemapatch:manifests="./helm/v1beta1" paths="./helm/v1beta1" output:dir="./helm/v1beta1"
'_output/tools/bin/controller-gen' schemapatch:manifests="./console/v1" paths="./console/v1" output:dir="./console/v1"
'_output/tools/bin/controller-gen' schemapatch:manifests="./console/v1alpha1" paths="./console/v1alpha1" output:dir="./console/v1alpha1"
'_output/tools/bin/controller-gen' schemapatch:manifests="./imageregistry/v1" paths="./imageregistry/v1" output:dir="./imageregistry/v1"
'_output/tools/bin/controller-gen' schemapatch:manifests="./operator/v1" paths="./operator/v1" output:dir="./operator/v1"
/home/christoph/go/src/github.com/openshift/api/operator/v1/types_console.go:198:2: must apply minlength to a string
Error: not all generators ran successfully
run `controller-gen schemapatch:manifests=./operator/v1 paths=./operator/v1 output:dir=./operator/v1 -w` to see all available markers, or `controller-gen schemapatch:manifests=./operator/v1 paths=./operator/v1 output:dir=./operator/v1 -h` for usage
make: *** [Makefile:29: update-codegen-crds-operator] Error 1
```

`// +kubebuilder:validation:UniqueItems=true` runs much longer, added a `uniqueItems: true` option to a yaml file, but fails when it validate it again:

```
$ make update build verify

...
The CustomResourceDefinition "consoles.operator.openshift.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[customization].properties[addPage].properties[disabledActions].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic
...
I0415 12:23:51.784975  535857 clientconn.go:897] blockingPicker: the picked transport is not ready, loop back to repick
W0415 12:23:51.784991  535857 clientconn.go:1223] grpc: addrConn.createTransport failed to connect to {http://127.0.0.1:2379  <nil> 0 <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:2379: connect: connection refused
make: *** [Makefile:50: verify-scripts] Error 1
```

**Comment by jhadvig** [comment]
@jerolimov then lets just keep `MinItems=1` and it should be ok :)

**Review by jhadvig — APPROVED** [review]
LGTM

**Comment by jhadvig** [comment]
@spadgett could you please review and approve the PR ? Thanks

**Comment by makambalaji** [comment]
This PR is related to Backend, So nothing to be tested as part of QE process. so adding the below label
/label qe-approved

**Comment by jerolimov** [comment]
Since the epic acceptance criteria of [ODC-5013](https://issues.redhat.com/browse/ODC-5013) have not changed and the epic has an `px-ack`, we approve this on behalf of PX.
/label px-approved

**Comment by spadgett** [comment]
> This PR is related to Backend, So nothing to be tested as part of QE process. so adding the below label

/hold

We shouldn't merge any APIs or backend changes until the feature has been tested. This might require testing the API, operator, backend, and frontend PRs together with cluster bot.

**Review by jhadvig — APPROVED** [review]
/lgtm

**Comment by gajanan-more** [comment]
Screenshot from 2021-04-27 14-48-06
Screenshot from 2021-04-27 14-46-45
Screenshot from 2021-04-27 14-46-40

I have reviewed the PR with the help of steps provided in the PR description. It is working as expected. There are couple of obervations: 
1. Invalid entry is ignored and it does not affect the Add page
2. You can not update the console-config YAML. If you want to change something, change it to cluster YAML and then it will get redirected to console-config YAML. 

/label qe-approved

File: operator/v1/types_console.go
Comment (spadgett) [inline operator/v1/types_console.go:193]: Is there any way we can list the IDs somewhere? Either here or in the docs? As an admin changing this, I wouldn't know what to set.

@christianvogt Thoughts?

Comment (jerolimov) [reply operator/v1/types_console.go:193]: @spadgett We provide a snippet for this in already merged PR https://github.com/openshift/console/pull/8717

Recording from the other PR:

Peek 2021-04-21 00-21

(fyi: The insertYAML automatically drops the comments.)

**Comment by jerolimov** [comment]
@spadgett this needs your approval so that we can bump it in console-operator. As you can see above the QE verified this on a cluster end to end. The only delay between the api, console-operator and console PRs will be bumps, rebase and approvals. Everything else is ready.

File: operator/v1/types_console.go
Comment (christianvogt) [reply operator/v1/types_console.go:193]: We don't want to put the IDs here right now because the IDs are dynamic.
The snippet that is shown above is dynamically generated.

**Review by spadgett — APPROVED** [review]
/approve

**Pull request merged**
